File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/dto/Dtos.kt
package com.hellotimepro.ktor.dto

import kotlinx.serialization.Serializable

// ── 统一响应外壳 ───────────────────────────────────────────────────────────

/** 成功响应外壳：`{ success, data, message, errorCode }`。 */
@Serializable
data class Envelope<T>(
    val success: Boolean,
    val data: T?,
    val message: String? = null,
    val errorCode: String? = null,
) {
    companion object {
        fun <T> ok(data: T): Envelope<T> = Envelope(true, data, null, null)
    }
}

@Serializable
data class ErrorDetail(val field: String, val message: String)

/** 失败响应外壳。data 恒为 null；details 仅 VALIDATION_ERROR / 带字段的 CONFLICT 时为数组。 */
@Serializable
data class ErrorEnvelope(
    val success: Boolean = false,
    val data: String? = null,
    val message: String,
    val errorCode: String,
    val details: List<ErrorDetail>? = null,
)

@Serializable
data class Pagination(val page: Int, val pageSize: Int, val total: Long, val totalPages: Int)

@Serializable
data class Paginated<T>(val items: List<T>, val pagination: Pagination)

// ── 用户 / 头像 ────────────────────────────────────────────────────────────

@Serializable
data class UserDto(
    val id: String,
    val email: String,
    val nickname: String,
    val avatarId: String,
    val createdAt: String,
)

@Serializable
data class UserBrief(val nickname: String, val avatarId: String)

@Serializable
data class AvatarDto(
    val id: String,
    val name: String,
    val primaryColor: String,
    val svgUrl: String? = null,
)

// ── 鉴权 ───────────────────────────────────────────────────────────────────

@Serializable
data class AuthTokens(
    val accessToken: String,
    val refreshToken: String,
    val accessTokenExpiresIn: Long,
    val refreshTokenExpiresIn: Long,
    val user: UserDto,
)

@Serializable
data class RegisterRequest(
    val email: String? = null,
    val password: String? = null,
    val nickname: String? = null,
    val avatarId: String? = null,
)

@Serializable
data class LoginRequest(val email: String? = null, val password: String? = null)

@Serializable
data class RefreshRequest(val refreshToken: String? = null)

@Serializable
data class LogoutRequest(val refreshToken: String? = null)

@Serializable
data class UpdateProfileRequest(val nickname: String? = null, val avatarId: String? = null)

@Serializable
data class ChangePasswordRequest(val currentPassword: String? = null, val newPassword: String? = null)

// ── 胶囊 ───────────────────────────────────────────────────────────────────

@Serializable
data class CreateCapsuleRequest(
    val title: String? = null,
    val content: String? = null,
    val openAt: String? = null,
    val inPlaza: Boolean? = null,
)

@Serializable
data class FavoriteRequest(val capsuleId: String? = null)

/** 单个胶囊详情：含 content（未开启为 null）。 */
@Serializable
data class CapsuleDetail(
    val id: String,
    val code: String,
    val title: String,
    val creator: UserBrief,
    val openAt: String,
    val createdAt: String,
    val inPlaza: Boolean,
    val favoriteCount: Int,
    val isOpened: Boolean,
    val content: String?,
    val favoritedByMe: Boolean,
)

/** 列表项：不含 content，含 contentPreview / favoritedAt。 */
@Serializable
data class CapsuleListItem(
    val id: String,
    val code: String,
    val title: String,
    val creator: UserBrief,
    val openAt: String,
    val createdAt: String,
    val inPlaza: Boolean,
    val favoriteCount: Int,
    val isOpened: Boolean,
    val favoritedByMe: Boolean,
    val favoritedAt: String? = null,
    val contentPreview: String? = null,
)

@Serializable
data class FavoriteResult(val capsuleId: String, val favoriteCount: Int, val favoritedAt: String)

// ── 健康检查 ───────────────────────────────────────────────────────────────

@Serializable
data class StackItem(val role: String, val name: String, val version: String, val iconUrl: String? = null)

@Serializable
data class StackInfo(val kind: String, val summary: String, val items: List<StackItem>)

@Serializable
data class HealthData(
    val status: String,
    val service: String,
    val version: String,
    val uptimeSeconds: Long,
    val stack: StackInfo,
)

// ── AI 建议 / 推荐 ─────────────────────────────────────────────────────────

@Serializable
data class CapsuleSuggestionRequest(val title: String? = null, val locale: String? = null)

@Serializable
data class CapsuleSuggestion(
    // 仅空标题模式下返回非 null，供前端回填。
    val title: String?,
    val content: String,
    val openInDays: Int,
    val openAt: String,
    val generatedBy: String,
    val cached: Boolean,
)

@Serializable
data class CapsuleRecommendationItem(val title: String, val hint: String, val openInDays: Int)

@Serializable
data class CapsuleRecommendationList(
    val items: List<CapsuleRecommendationItem>,
    val generatedBy: String,
    val cached: Boolean,
)

File: backends/aspnet/src/Dto/Dtos.cs
using System.Text.Json.Serialization;

namespace HelloTimePro.Aspnet.Dto;

// ── 统一响应外壳 ───────────────────────────────────────────────────────────

/// <summary>成功响应外壳：<c>{ success, data, message, errorCode }</c>。</summary>
public sealed record Envelope<T>(bool Success, T? Data, string? Message, string? ErrorCode)
{
    public static Envelope<T> Ok(T data) => new(true, data, null, null);
}

public sealed record ErrorDetail(string Field, string Message);

/// <summary>失败响应外壳。data 恒为 null；details 仅 VALIDATION_ERROR / 带字段的 CONFLICT 时出现。</summary>
public sealed record ErrorEnvelope(string Message, string ErrorCode)
{
    public bool Success => false;
    public object? Data => null;

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public IReadOnlyList<ErrorDetail>? Details { get; init; }
}

public sealed record Pagination(int Page, int PageSize, long Total, int TotalPages);

public sealed record Paginated<T>(IReadOnlyList<T> Items, Pagination Pagination);

// ── 用户 / 头像 ────────────────────────────────────────────────────────────

public sealed record UserDto(string Id, string Email, string Nickname, string AvatarId, string CreatedAt);

public sealed record UserBrief(string Nickname, string AvatarId);

public sealed record AvatarDto(string Id, string Name, string PrimaryColor, string? SvgUrl);

// ── 鉴权 ───────────────────────────────────────────────────────────────────

public sealed record AuthTokens(
    string AccessToken,
    string RefreshToken,
    long AccessTokenExpiresIn,
    long RefreshTokenExpiresIn,
    UserDto User);

public sealed record RegisterRequest(string? Email, string? Password, string? Nickname, string? AvatarId);

public sealed record LoginRequest(string? Email, string? Password);

public sealed record RefreshRequest(string? RefreshToken);

public sealed record LogoutRequest(string? RefreshToken);

public sealed record UpdateProfileRequest(string? Nickname, string? AvatarId);

public sealed record ChangePasswordRequest(string? CurrentPassword, string? NewPassword);

// ── 胶囊 ───────────────────────────────────────────────────────────────────

public sealed record CreateCapsuleRequest(string? Title, string? Content, string? OpenAt, bool? InPlaza);

public sealed record FavoriteRequest(string? CapsuleId);

/// <summary>单个胶囊详情：含 content（未开启为 null）。</summary>
public sealed record CapsuleDetail(
    string Id,
    string Code,
    string Title,
    UserBrief Creator,
    string OpenAt,
    string CreatedAt,
    bool InPlaza,
    int FavoriteCount,
    bool IsOpened,
    string? Content,
    bool FavoritedByMe);

/// <summary>列表项：不含 content，含 contentPreview / favoritedAt。</summary>
public sealed record CapsuleListItem(
    string Id,
    string Code,
    string Title,
    UserBrief Creator,
    string OpenAt,
    string CreatedAt,
    bool InPlaza,
    int FavoriteCount,
    bool IsOpened,
    bool FavoritedByMe,
    string? FavoritedAt,
    string? ContentPreview);

public sealed record FavoriteResult(string CapsuleId, int FavoriteCount, string FavoritedAt);

// ── 健康检查 ───────────────────────────────────────────────────────────────

public sealed record StackItem(string Role, string Name, string Version, string? IconUrl);

public sealed record StackInfo(string Kind, string Summary, IReadOnlyList<StackItem> Items);

public sealed record HealthData(
    string Status,
    string Service,
    string Version,
    long UptimeSeconds,
    StackInfo Stack);

// ── AI 建议 / 推荐 ─────────────────────────────────────────────────────────

public sealed record CapsuleSuggestionRequest(string? Title, string? Locale);

public sealed record CapsuleSuggestion(
    // 仅空标题模式下返回非 null，供前端回填。
    string? Title,
    string Content,
    int OpenInDays,
    string OpenAt,
    string GeneratedBy,
    bool Cached);

public sealed record CapsuleRecommendationItem(string Title, string Hint, int OpenInDays);

public sealed record CapsuleRecommendationList(
    IReadOnlyList<CapsuleRecommendationItem> Items,
    string GeneratedBy,
    bool Cached);
